hogben/utils.py

```diff
@@ -163,77 +157,182 @@ def fisher(qs: list[np.ndarray],
-                             for q, model in list(zip(qs, models))])
-
-        parameter.value = old  # Reset the parameter.
-
-        J[:, i] = (y2 - y1) / (x2 - x1)  # Calculate the gradient.
-
-    # Calculate the reflectance for each model for the given Q values.
-    r = np.concatenate([reflectivity(q, model)
-                        for q, model in list(zip(qs, models))])
-
-    # Calculate the Fisher information matrix using equations from the paper.
-    M = np.diag(np.concatenate(counts) / r, k=0)
-    g = np.dot(np.dot(J.T, M), J)
-
-    # If there are multiple parameters,
-    # scale each parameter's information by its "importance".
-    if len(xi) > 1:
-        if isinstance(xi[0], refnx.analysis.Parameter):
-            lb = np.array([param.bounds.lb for param in xi])
-            ub = np.array([param.bounds.ub for param in xi])
-
-        elif isinstance(xi[0], bumps.parameter.Parameter):
-            lb = np.array([param.bounds.limits[0] for param in xi])
-            ub = np.array([param.bounds.limits[1] for param in xi])
-
-        # Scale each parameter with their specified importance,
-        # scale with one if no importance was specified.
-        importance_array = []
-        for param in xi:
-            if hasattr(param, "importance"):
-                importance_array.append(param.importance)
-            else:
-                importance_array.append(1)
-        importance = np.diag(importance_array)
+
+    def __init__(self,
+                 qs: list[np.ndarray],
+                 xi: list[Union['refnx.analysis.Parameter',
+                                'bumps.parameter.Parameter']],
+                 counts: list[int],
+                 models: list[Union['refnx.reflect.ReflectModel',
+                                    'refl1d.experiment.Experiment']],
+                 step: float = 0.005):
+        """Constructor for the Fisher matrix class.
+
+        Args:
+            qs: The Q points for each model.
+            xi: The varying model parameters.
+            counts: incident neutron counts corresponding to each Q value.
+            models: models to calculate gradients with.
+            step: step size to take when calculating gradient.
+            g: The Fisher information matrix
+            min_eigenval: The minimum eigenvalue associated with the Fisher
+             information matrix
+        """
+        self.qs = qs
+        self.xi = xi
+        self.counts = counts
+        self.models = models
+        self.step = step
+
+    @property
+    def n(self) -> int:
+        """The total amount of datapoints.
+
+        Returns:
+            int: total amount of datapoints.
+        """
+        return sum(len(q) for q in self.qs)
+
+    @property
+    def m(self) -> int:
+        """The total amount of parameters.
+
+        Returns:
+            int: total amount of parameters.
+        """
+        return len(self.xi)
+
+    @property
+    def g(self) -> np.ndarray:
```

Comment:
`information_matrix` may be a better name, not 10% sure, but it is clearer for someone who hasn't read the derivation paper etc

Reply:
Yeah I think it would be, at least for the user-facing side.